pkg/app/api/analysisresultstore/store.go

```diff
@@ -0,0 +1,84 @@
+// Copyright 2021 The PipeCD Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package analysisresultstore
+
+import (
+	"context"
+	"errors"
+
+	"go.uber.org/zap"
+
+	"github.com/pipe-cd/pipe/pkg/cache"
+	"github.com/pipe-cd/pipe/pkg/filestore"
+	"github.com/pipe-cd/pipe/pkg/model"
+)
+
+type Store interface {
+	// GetMostRecentSuccessfulAnalysisMetadata gives back the most recent successful analysis metadata of the specified application.
+	GetMostRecentSuccessfulAnalysisMetadata(ctx context.Context, applicationID string) (*model.AnalysisMetadata, error)
+	// PutStateSnapshot updates the most recent successful analysis metadata of the specified application.
+	PutMostRecentSuccessfulAnalysisMetadata(ctx context.Context, applicationID string, snapshot *model.AnalysisMetadata) error
+}
+
+type store struct {
+	backend *analysisFileStore
+	cache   *analysisCache
+	logger  *zap.Logger
+}
+
+func NewStore(fs filestore.Store, c cache.Cache, logger *zap.Logger) Store {
+	return &store{
+		backend: &analysisFileStore{
+			backend: fs,
+		},
+		cache: &analysisCache{
+			backend: c,
+		},
+		logger: logger.Named("latest-analysis-store"),
+	}
+}
+
+func (s *store) GetMostRecentSuccessfulAnalysisMetadata(ctx context.Context, applicationID string) (*model.AnalysisMetadata, error) {
+	cacheResp, err := s.cache.Get(applicationID)
+	if err != nil && !errors.Is(err, cache.ErrNotFound) {
+		s.logger.Error("failed to get the most recent successful analysis metadata from cache", zap.Error(err))
+	}
+	if cacheResp != nil {
+		return cacheResp, nil
+	}
```

Comment:
I think maybe we should not use the cache layer for this. Here are some reasons:
- The file object is containing the latest analysis metadata, this means caching could cause returning stale data.
- At the same time only one deployment of an application could be executed so this API is not called much.

Reply:
That is so true. To avoid returning stale data, I [didn't let piped's analysis store cache periodically](https://github.com/pipe-cd/pipe/blob/analysis-store/pkg/app/piped/apistore/analysisresultstore/store.go). Let me remove this.